### PR TITLE
Clean up child compilation chunk graph to avoid memory leak

### PR DESCRIPTION
**Pull request by helloitsjoe**
Fixes https://github.com/webpack/webpack/issues/17851

Cleans up child compilation chunk graph which includes references to the compiler and parent compiler, causing a memory leak. It looks to me like child compilers were accidentally overlooked when this change was made to clean up modules/chunks in the parent:  https://github.com/webpack/webpack/pull/12894/files#diff-732395f8fcd972919381425c80194bb0def0a2af50c0940762299605a81c0006R360-R372

Details in the reproduction repo here: https://github.com/helloitsjoe/webpack-memory-leak (let me know if you want me to add more detail inline in this PR description)

Heap diff before | Heap diff after
--- | ---
Heap diff before | Heap diff after

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 0f517fd</samp>

Refactor `Compiler` class to clear chunk graph data of child compilations in `close` method. This prevents memory leaks and improves performance.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 0f517fd</samp>

* Clear chunk graph data for child compilation chunks to prevent memory leaks and improve performance ([link](https://github.com/webpack/webpack/pull/17853/files?diff=unified&w=0#diff-732395f8fcd972919381425c80194bb0def0a2af50c0940762299605a81c0006R384-R389))

### CODE REVIEW

**Comment by helloitsjoe** [comment]
I plan on adding tests after I familiarize myself with Webpack's unit test structure, I'll leave this PR in draft until then.

File: lib/Compiler.js [hunk 1]
```diff
@@ -381,6 +381,12 @@ class Compiler {
 	// e.g. move compilation specific info from Modules into ModuleGraph
 	_cleanupLastCompilation() {
 		if (this._lastCompilation !== undefined) {
+			for (const childCompilation of this._lastCompilation.children) {
+				for (const chunk of childCompilation.chunks) {
+					ChunkGraph.clearChunkGraphForChunk(chunk);
```
Comment (helloitsjoe) [inline lib/Compiler.js:386]: I was initially also up modules as well:

```js
  for (const module of childCompilation.modules) {
				ChunkGraph.clearChunkGraphForModule(module);
				ModuleGraph.clearModuleGraphForModule(module);
				module.cleanupForCache();
  }
```

but it looks like `ChunkGraph.clearChunkGraphForChunk(chunk)` is the only change necessary to fix the memory leak.

**Comment by helloitsjoe** [comment]
I noticed the original PR to add cleanup doesn't include tests, and I'm not exactly sure how to go about writing tests for this case. If you would like me to add tests just let me know and I'll look further into how I could test this!

**Review by alexander-akait — APPROVED** [review]
Good catch, I am fine with it without tests, because we forgot to cleanup child compilations

**Comment by alexander-akait** [comment]
Thank you, release will be soon

**Pull request merged**